torch/distributed/_functional_collectives.py

```diff
@@ -704,20 +704,21 @@ def _all_gather_into_tensor_coalesced_fallback(output_tensors, input_tensors, gr
     # NCCL's PG::all_gather with multiple tensors is broken, it only works for the multi-device setting
     #  and fails if you mix same-size with different-size tensor lists.
     # _coalescing_manager crashed NCCL when used with all_gather_into_tensor.
-    work_list = []
-    if input_tensors[0].is_cpu:
+    if input_tensors[0].is_cpu or not async_op:
+        work_list = []
         out_tensors_sliced = [
             list(torch.chunk(out_tensor, dist.get_world_size(group)))
             for out_tensor in output_tensors
         ]
         for shard, out_tensor in zip(input_tensors, out_tensors_sliced):
             work = c10d.all_gather(out_tensor, shard, group=group, async_op=async_op)
             work_list.append(work)
+        return work_list
     else:
-        for shard, out_tensor in zip(input_tensors, output_tensors):
-            work = c10d.all_gather_into_tensor(out_tensor, shard, group=group, async_op=async_op)
-            work_list.append(work)
-    return work_list
+        with c10d._coalescing_manager(group=group, async_ops=True) as cm:
+            for in_t, out_t in zip(input_tensors, output_tensors):
+                dist.all_gather_into_tensor(out_t, in_t, group=group, async_op=True)
+        return cm
```

Comment:
hmmm I think we should return the work/asyncCollectiveTensor list instead of `cm` for the nccl path?

Reply:
Actually no, when using async collectives with _CoalescingManager, we need to call wait on it as the individual collectives won't return a work object.